File: scripts/subscriber.py
#!/usr/bin/env python
import rospy
from performance_tests.msg import SuperAwesome

time1 = 0
sumTime = 0
nb = 0
firstUsage = True
rep = "additionalFiles/"

def writeToFile(sumToWrite, rate, serverType):
    global rep
    fileName =  serverType + "ToPython" 
    #rospy.loginfo(rospy.get_caller_id() +" fileName = %s", fileName)
    
    myfile = open(fileName, "a")
    myfile.write(str(rate)+"|"+str(sumToWrite)+"\n")
    myfile.close()



def callback(data):

    global firstUsage
    global sumTime
    global nb
    global time1

    
    time2 = rospy.get_time()
    #rospy.loginfo(rospy.get_caller_id() + "I heard %s", data.s)
    if "FIN" in data.s:
        #rospy.loginfo(rospy.get_caller_id() + " ending process")
        serverInfo = data.s.split("|")
        writeToFile((sumTime/nb),serverInfo[2],serverInfo[1])
        return
        

    if firstUsage:
        firstUsage = False
    else:
        sumTime += time2 - time1;
        nb+=1;
        #rospy.loginfo("curent time average = %f s", (sumTime/nb));
    time1 = rospy.get_time()
    
    
def listener():
    rospy.init_node('listener', anonymous=True)
    rospy.Subscriber("SuperTopic", SuperAwesome, callback)
    rospy.spin()

if __name__ == '__main__':
    listener()

File: src/subscriber.cpp
#include "ros/ros.h"
#include <performance_tests/SuperAwesome.h>
#include <string>
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

double time1, sum = 0;
int nb = 0;
bool firstUsage = true;


template<typename Out>
void split(const std::string &s, char delim, Out result) {
    std::stringstream ss;
    ss.str(s);
    std::string item;
    while (std::getline(ss, item, delim)) {
        *(result++) = item;
    }
}


std::vector<std::string> split(const std::string &s, char delim) {
    std::vector<std::string> elems;
    split(s, delim, std::back_inserter(elems));
    return elems;
}


void writeToFile(double sumToWrite, std::string rate, std::string serverType)
{
  std::string fileName = serverType+"ToCpp";
  //ROS_INFO_STREAM("fileName = " << fileName);
    
  ofstream myfile;
  myfile.open (fileName.c_str(), ios::app);
    
  std::ostringstream sstream;
  sstream << rate << "|" << sumToWrite;
  std::string varAsString = sstream.str();
  myfile << varAsString+"\n";
  myfile.close();
}

void pCallback(const performance_tests::SuperAwesome::ConstPtr& msg)
{
  double time2 = ros::Time::now().toSec();
  
  if (msg->s.find("FIN") != std::string::npos)
  {
    //ROS_INFO_STREAM("ending process");
    std::vector<std::string> serverInfo = split(msg->s.c_str(),'|');
    writeToFile((sum/nb),serverInfo[2],serverInfo[1]);
    return;
  }
  

  //ROS_INFO("I heard: [%s]", msg->s.c_str());
  if (firstUsage)
    firstUsage = false;
  else
  {
    sum += time2 - time1;
    nb++;
    //ROS_INFO_STREAM("curent time average = " << (sum/nb) << " s" );
  }
  time1 = ros::Time::now().toSec();
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "subs");
  ros::NodeHandle n;
  ros::Subscriber sub = n.subscribe("SuperTopic", 1000, pCallback);
  ros::spin();

  return 0;
}
